### ISSUE TITLE
error

### ISSUE THREAD

**Issue by liaguiro** [NBISweden/GUESSmyLT#17]
Hello,

I'm getting the following error after sending this job:
GUESSmyLT --reads ~/sample1.fastq  --reference ~/referenceGenome.fa --mode genome --mapped sample1_bowtie2.coordSorted.bam --organism euk

Reference and bam have Sequence ID matching. We can continue.
Traceback (most recent call last):
  File "~/.conda/envs/py3Env/bin/GUESSmyLT", line 8, in <module>
    sys.exit(main())
  File "~/.conda/envs/py3Env/lib/python3.6/site-packages/GUESSmyLT/GUESSmyLT.py", line 305, in main
    if get_type(args.mapped) == "single":
  File "~/.conda/envs/py3Env/lib/python3.6/site-packages/GUESSmyLT/GUESSmyLT.py", line 94, in get_type
    nb_reads_paired = int(str(out.decode("utf-8")).split("\n")[0])
ValueError: invalid literal for int() with base 10: ''

Any suggestion on how to fix this error?

Thanks

**Comment by Juke34** [comment]
The samtools command "samtools view -c -f 1 sample1.fastq" do not return what is expected. Could you run this command in a terminal and give me the result? Could you tell me the samtools version too?

**Comment by liaguiro** [comment]
Hello,

I had to uninstall the samtools that was installed when installing guessmylt in my conda environment. Samtools worked after loading the module (module load samtools/1.11).
samtools view -c -f 1 sample1.bam
0

When I run guessmylt, I'm getting the following error:

GUESSmyLT --reads ~/sample1.fastq  --reference ~/referenceGenome.fa --mode genome --mapped sample1.bam --organism euk
Reference and bam have Sequence ID matching. We can continue.
Building DAG of jobs...
Error: Directory cannot be locked. Please make sure that no other Snakemake process is trying to create the same files in the following directory:
~/
If you are sure that no other instances of snakemake are running on this directory, the remaining lock was likely caused by a kill signal or a power loss. It can be removed with the --unlock argument.

I added the --unlock option, but I got the following error:
usage: GUESSmyLT [-h] [--organism ORGANISM] [--reads READS [READS ...]]
                 [--subsample SUBSAMPLE] [--reference REFERENCE] [--mode MODE]
                 [--annotation ANNOTATION] [--mapped MAPPED]
                 [--threads THREADS] [--memory MEMORY] [--output OUTPUT] [-n]
GUESSmyLT: error: unrecognized arguments: --unlock

I don't know which package is causing the error. Any suggestion on how to fix this error?

**Comment by Juke34** [comment]
This is now a snakemake problem. You can remove the ".snakemake" directory and start again.

**Comment by liaguiro** [comment]
Hi,
I'm getting this output on my terminal, and it stays there. I increased number of cores and memory, but it's still showing the message.
Does this step take a long time to finish? Or Is the program stuck? If it is, how can I make it continue working?

Thanks.

Building DAG of jobs...
Using shell: /usr/bin/bash
Provided cores: 8
Rules claiming more threads will be scaled down.
Job counts:
	count	jobs
	1	bowtie2_index
	1	bowtie2_map
	1	busco_ann
	1	check_programs
	1	inferr
	1	samtools_sort_new_mapped
	1	subsample_fastq
	7

[Wed May  5 08:37:05 2021]
rule check_programs:
    output: programs.checked
    jobid: 4
    priority: 2

Job counts:
	count	jobs
	1	check_programs
	1
Touching output file programs.checked.
[Wed May  5 08:37:06 2021]
Finished job 4.
1 of 7 steps (14%) done

[Wed May  5 08:37:06 2021]
rule subsample_fastq:
    input: /sample1.fastq, programs.checked
    output: intermediate_data/sample1.full.fastq.gz
    jobid: 1
    priority: 1

Job counts:
	count	jobs
	1	subsample_fastq
	1

**Comment by Juke34** [comment]
Subsample is slow because GUESSmyLT use python to perform this task, it would much quicker by using seqtk.
If you want to check it is running try `top` or `htop`and look for the corresponding task.
To make it faster you can use a lower value for the subsampling.

**Comment by liaguiro** [comment]
Adding the subsample option helped, but I got a new error now (see below). I'm not sure how to fix it. Any suggestions? (FYI, I'm new at bioinformatics)

Job counts:
	count	jobs
	1	subsample_fastq
	1
Subsampling 500000 reads of /sample1.fastq to new read file: /sample1.500000.fastq.gz
cat: write error: Broken pipe
Reformating read headers in fastq file: /sample1.500000.fastq.gz
[Wed May  5 13:49:47 2021]
Finished job 1.
2 of 7 steps (29%) done

[Wed May  5 13:49:47 2021]
rule bowtie2_index:
    input: /referenceGenome.fa
    output: intermediate_data/index/referenceGenome.1.bt2, intermediate_data/index/referenceGenome.2.bt2, intermediate_data/index/referenceGenome.3.bt2, intermediate_data/index/referenceGenome.4.bt2, intermediate_data/index/referenceGenome.rev.1.bt2, intermediate_data/index/referenceGenome.rev.2.bt2
    log: logs/bowtie2_index_referenceGenome.log
    jobid: 6
    threads: 8

Job counts:
	count	jobs
	1	bowtie2_index
	1
/.conda/envs/py3Env/bin/bowtie2-build-s: symbol lookup error: /.conda/envs/py3Env/bin/bowtie2-build-s: undefined symbol: _ZN3tbb10interface78internal15task_arena_base19internal_initializeEv
[Wed May  5 13:49:48 2021]
Error in rule bowtie2_index:
    jobid: 0
    output: intermediate_data/index/referenceGenome.1.bt2, intermediate_data/index/referenceGenome.2.bt2, intermediate_data/index/referenceGenome.3.bt2, intermediate_data/index/referenceGenome.4.bt2, intermediate_data/index/referenceGenome.rev.1.bt2, intermediate_data/index/referenceGenome.rev.2.bt2
    log: logs/bowtie2_index_referenceGenome.log (check log file(s) for error message)

RuleException:
CalledProcessError in line 42 of /.conda/envs/py3Env/lib/python3.6/site-packages/GUESSmyLT/bowtie2:
Command 'set -euo pipefail;  bowtie2-build /referenceGenome.fa intermediate_data/index/referenceGenome | tee logs/bowtie2_index_referenceGenome.log' returned non-zero exit status 127.
  File "/.conda/envs/py3Env/lib/python3.6/site-packages/GUESSmyLT/bowtie2", line 42, in __rule_bowtie2_index
  File "/.conda/envs/py3Env/lib/python3.6/concurrent/futures/thread.py", line 56, in run
Exiting because a job execution failed. Look above for error message
Shutting down, this might take some time.
Exiting because a job execution failed. Look above for error message
Complete log: /.snakemake/log/2021-05-05T134920.742698.snakemake.log

**Comment by Juke34** [comment]
Might  be due to bowtie version used...
You migh try to run the command yourself  `bowtie2-build /referenceGenome.fa intermediate_data/index/referenceGenome | tee logs/bowtie2_index_referenceGenome.log` from the proper location to see the real error. What type of  machine are you running the tool? What is the bowtie2 version?

**Comment by liaguiro** [comment]
Hi,
I'm using bowtie2 v.2.3.5 in my conda environment, I'm working on a cluster (I think the specification is linus/centos/7.x/x86_64), and I got the same error after running bowtie2-build in conda:
/.conda/envs/py3Env/bin/bowtie2-build-s: symbol lookup error: /.conda/envs/py3Env/bin/bowtie2-build-s: undefined symbol: _ZN3tbb10interface78internal15task_arena_base19internal_initializeEv

It worked after removing bowtie2 from my conda environment and loaded the bowtie2 available in the cluster. Also, I had to load busco and augustus.

Now, I'm getting an error from samtools (see below). I think, I need to use the samtools that you used when you developed this package because they removed -f option from samtools v.1.11 (this is the version that I'm using). Which samtools version works with guessmylt? (FYI, I had to uninstall samtools from my conda too and load samtools available in the cluster)

[Thu May  6 10:44:04 2021]
Finished job 5.
4 of 7 steps (57%) done

[Thu May  6 10:44:04 2021]
rule samtools_sort_new_mapped:
    input: intermediate_data/sample1_on_ref_referenceGenome_mapped.bam
    output: intermediate_data/from_raw_sample1500000_sorted.bam
    log: logs/samtools_sort_new_mapped_sample1_on_ref_referenceGenome.log
    jobid: 3
    threads: 8

Job counts:
	count	jobs
	1	samtools_sort_new_mapped
	1
out:1.11

samtools version being used:1.11
sort: invalid option -- 'f'
Usage: samtools sort [options...] [in.bam]
Options:
  -l INT     Set compression level, from 0 (uncompressed) to 9 (best)
  -u         Output uncompressed data (equivalent to -l 0)
  -m INT     Set maximum memory per thread; suffix K/M/G recognized [768M]
  -M         Use minimiser for clustering unaligned/unplaced reads
  -K INT     Kmer size to use for minimiser [20]
  -n         Sort by read name (not compatible with samtools index command)
  -t TAG     Sort by value of TAG. Uses position as secondary index (or read name if -n is set)
  -o FILE    Write final output to FILE rather than standard output
  -T PREFIX  Write temporary files to PREFIX.nnnn.bam
  --no-PG    do not add a PG line
      --input-fmt-option OPT[=VAL]
               Specify a single input file format option in the form
               of OPTION or OPTION=VALUE
  -O, --output-fmt FORMAT[,OPT[=VAL]]...
               Specify output format (SAM, BAM, CRAM)
      --output-fmt-option OPT[=VAL]
               Specify a single output file format option in the form
               of OPTION or OPTION=VALUE
      --reference FILE
               Reference sequence FASTA FILE [null]
  -@, --threads INT
               Number of additional threads to use [0]
      --verbosity INT
               Set level of verbosity
[Thu May  6 10:44:05 2021]
Error in rule samtools_sort_new_mapped:
    jobid: 0
    output: intermediate_data/from_raw_sample1500000_sorted.bam
    log: logs/samtools_sort_new_mapped_sample1_on_ref_referenceGenome.log (check log file(s) for error message)

RuleException:
CalledProcessError in line 62 of /.conda/envs/py3Env/lib/python3.6/site-packages/GUESSmyLT/samtools:
Command 'set -euo pipefail;  samtools sort intermediate_data/sample1_on_ref_referenceGenome_mapped.bam -f intermediate_data/from_raw_sample1500000_sorted.bam
         samtools index intermediate_data/from_raw_sample1500000_sorted.bam
         rm intermediate_data/sample1_on_ref_referenceGenome_mapped.bam' returned non-zero exit status 1.
  File "/.conda/envs/py3Env/lib/python3.6/site-packages/GUESSmyLT/samtools", line 62, in __rule_samtools_sort_new_mapped
  File "/.conda/envs/py3Env/lib/python3.6/concurrent/futures/thread.py", line 56, in run
Exiting because a job execution failed. Look above for error message
Shutting down, this might take some time.
Exiting because a job execution failed. Look above for error message
Complete log: /.snakemake/log/2021-05-06T095200.359097.snakemake.log

**Comment by Juke34** [comment]
GUESSmyLT deals on the fly the samtools command depending the version. If version is < 1.3 we use the -f flag to define the output file otherwise we redirect the stdout with ">".
So apparently the version 1.11 is seen as inferior to 1.3... This is a bug

**Comment by Juke34** [comment]
I have fixed it in branch `testversion` 
Follow the instruction here to install this version:
https://github.com/NBISweden/GUESSmyLT/tree/testversion#installation-with-git
but move to the `testversion` branch after the following command

```
git clone https://github.com/NBISweden/GUESSmyLT.git
cd GUESSmyLT/
```

by doing 
```
git checkout testversion
```

then continue the instructions.

**Comment by liaguiro** [comment]
Hi,

I installed the test version, but I got an error (looks like the same as before... looking for samtools sort)

[Mon May 10 16:22:06 2021]
rule samtools_sort_new_mapped:
    input: intermediate_data/sample1_on_ref_referenceGenome_mapped.bam
    output: intermediate_data/from_raw_sample1500000_sorted.bam
    log: logs/samtools_sort_new_mapped_sample1_on_ref_referenceGenome.log
    jobid: 3
    threads: 8

Job counts:
	count	jobs
	1	samtools_sort_new_mapped
	1
out:1.11

samtools version being used:1.11
sort: invalid option -- 'f'
Usage: samtools sort [options...] [in.bam]
Options:
  -l INT     Set compression level, from 0 (uncompressed) to 9 (best)
  -u         Output uncompressed data (equivalent to -l 0)
  -m INT     Set maximum memory per thread; suffix K/M/G recognized [768M]
  -M         Use minimiser for clustering unaligned/unplaced reads
  -K INT     Kmer size to use for minimiser [20]
  -n         Sort by read name (not compatible with samtools index command)
  -t TAG     Sort by value of TAG. Uses position as secondary index (or read name if -n is set)
  -o FILE    Write final output to FILE rather than standard output
  -T PREFIX  Write temporary files to PREFIX.nnnn.bam
  --no-PG    do not add a PG line
      --input-fmt-option OPT[=VAL]
               Specify a single input file format option in the form
               of OPTION or OPTION=VALUE
  -O, --output-fmt FORMAT[,OPT[=VAL]]...
               Specify output format (SAM, BAM, CRAM)
      --output-fmt-option OPT[=VAL]
               Specify a single output file format option in the form
               of OPTION or OPTION=VALUE
      --reference FILE
               Reference sequence FASTA FILE [null]
  -@, --threads INT
               Number of additional threads to use [0]
      --verbosity INT
               Set level of verbosity
[Mon May 10 16:22:07 2021]
Error in rule samtools_sort_new_mapped:
    jobid: 0
    output: intermediate_data/from_raw_sample1500000_sorted.bam
    log: logs/samtools_sort_new_mapped_sample1_on_ref_referenceGenome.log (check log file(s) for error message)

RuleException:
CalledProcessError in line 62 of /.conda/envs/py3Env/lib/python3.6/site-packages/GUESSmyLT/samtools:
Command 'set -euo pipefail;  samtools sort intermediate_data/sample1_on_ref_referenceGenome.bam -f intermediate_data/from_raw_referenceGenome500000_sorted.bam
         samtools index intermediate_data/from_raw_sample1500000_sorted.bam
         rm intermediate_data/sample1_on_ref_referenceGenome_mapped.bam' returned non-zero exit status 1.
  File "/.conda/envs/py3Env/lib/python3.6/site-packages/GUESSmyLT/samtools", line 62, in __rule_samtools_sort_new_mapped
  File "/.conda/envs/py3Env/lib/python3.6/concurrent/futures/thread.py", line 56, in run
Exiting because a job execution failed. Look above for error message
Shutting down, this might take some time.
Exiting because a job execution failed. Look above for error message
Complete log: /.snakemake/log/2021-05-10T153404.465872.snakemake.log
